### PR TITLE
Streamline the UI a bit

### PR DESCRIPTION
**Pull request by hoylemd**
refs #41 

While working on #40, (and, well, learning how 2 lua), I made some changes that I think improve the user and development experience.

I haven't run the faketorio tests yet, mainly because I don't know how, and didn't really have internet access while working on these, so I suppose I should probably do that too.

Anyways, run down of the changes:

Main changes:
- Localize the main todo button
- Add explicit styles for UI elements, and use them
- Change the 'show button when minimized' setting to just always show the button
- Main Todo List button now toggles the UI, just like the hotkey
- Unify add and edit UIs
- Move add/edit UI init into UI module

QOL:
- ~todo.log also logs to main log~ Removed because lint doesn't like it and I don't think I can change the travis config to allow it. Also, not really super helpful.

fixes:
- Add chinese localization entries for auto-assign (in english, but at least they're there?)
- squelch the 'unknown element' warning when a non-todo gui element is clicked.

See in-line github comments for some more detailed explanations. (still coming)

Screenshots!

Before:
screenshot 2018-08-06 16 04 39

After:
screenshot 2018-08-06 15 51 03

### CODE REVIEW

**Comment by JonasJurczok** [comment]
Hi,
thanks for the PR.

Regarding the faketorio tests.. ideally it should be as easy as running `luarocks install faketorio` and then a `faketorio test` in the mod folder. Unfortunately there seems to be a problem with installing faketorio on windows (I was doing all developement on Linux until recently and now have to get my win environment to work).

I like the new styles. This also fixed the line break problem. nice :)

I'm not sure about keeping the button there all the time. The initial idea was to save that screenspace to be able to show other buttons etc. But tbh I play with the button invisible anyway.

I'll definitely take a look at your changes.

**Comment by hoylemd** [comment]
I'm working on macos, actually.  I was able to get faketorio working, but it took some tinkering, and a lot of the functionality didn't really work (e.g. it couldn't create or find the new save when running `faketorio test`, but I just started a new game and ran `/faketorio` and it worked.  There's only supposed to be the one feature, right?

l updated the label for the button setting so that it just says 'show the button in the mod ui'.  That way, you can still play with the invisible button, but if you're like me, you can have it persist.  I see a lot of other mods do this too (helmod and outpost planner are 2 examples).  Also updated the name of the setting to reflect it's functionality.

I updated the french and german localizations too, but my french and german aren't *great*, so if they seem weird to a native speaker please suggest an improvement.  Thankfully, the longer descriptions seemed to already describe the functionality I was going for, so I only had to modify the short ones.

I'm also not 100% sure I updated the faketorio test correctly, and given that CI is failing, I'm thinking I didn't.  So I'll check that out next.

File: src/control.lua [hunk 1]
```diff
@@ -22,16 +22,15 @@ end)
...
     local player = game.players[event.player_index]
     local key = event.setting
     todo.on_runtime_mod_setting_changed(player, key)
-end)
\ No newline at end of file
+end)
+
+require("faketorio.core")
+
+require("faketorio.features.UI_feature")
```
Comment (hoylemd) [inline src/control.lua]: I don't know why this is missing on master now, I think it was in master when I started this branch?  Seems to be important for running faketorio tests ¯\_(ツ)_/¯

File: src/todo/UI.lua [hunk 2]
```diff
@@ -2,11 +2,14 @@
 function todo.create_minimized_button(player)
     todo.log("Creating Basic UI for player " .. player.name)
 
-    if not todo.get_maximize_button(player) and not todo.get_main_frame(player) and todo.show_minimized(player) then
+    if (not todo.get_maximize_button(player)
+      and not todo.get_main_frame(player)
+      and todo.show_button(player) ) then
```
Comment (hoylemd) [inline src/todo/UI.lua]: I thought this was easier to read, but maybe just me.

File: src/todo/UI.lua [hunk 3]
```diff
@@ -121,40 +135,47 @@ function todo.create_add_edit_frame(player)
...
         caption = {"todo.add_task"}
     })
 
-    local textbox = table.add({
+    table.add({
         type = "text-box",
-        name = "todo_new_task_textbox"
+        style = "todo_textbox_default",
+        name = "todo_new_task_textbox",
+        text = task and task.task or ""
```
Comment (hoylemd) [inline src/todo/UI.lua]: I don't know if this is common in lua, but this is sort of a 'poor man's ternary'.

If task is truthy, then the result of the expression will be task.task. Otherwise, emptystring. Analogous to

text = task ? task.task : ""; in javascript.

(if this is already a well-known pattern, sorry. I'm very new to lua :p)

I used this in a few places, not just here.

File: src/todo/UI.lua [hunk 4]
```diff
@@ -62,51 +68,59 @@ function todo.create_task_table(frame, player)
...
         caption = ""
     })
 
     return table
 end
 
-function todo.create_add_edit_frame(player)
+function todo.create_add_edit_frame(player, task)
     local gui = player.gui.center
+    task = task or nil
```
Comment (hoylemd) [inline src/todo/UI.lua]: Not sure if this is really necessary, but better safe than sorry, I guess?

File: src/todo/UI.lua [hunk 5]
```diff
@@ -121,40 +135,47 @@ function todo.create_add_edit_frame(player)
...
-    local players, _, c = todo.get_player_list()
-    local index = 1
-    if( todo.is_auto_assign(player) and c == 1 ) then
-        index = 2
+    local players, lookup, c = todo.get_player_list()
+    local assign_index = 1
+    if task then
+        assign_index = task.assignee and lookup[task.assignee] or 1
+    else
+        if( todo.is_auto_assign(player) and c == 1 ) then
```
Comment (hoylemd) [inline src/todo/UI.lua]: Hmm, would it be better to merge this with the `else` to form an `elseif`?

File: src/todo/todo.lua [hunk 6]
```diff
@@ -96,32 +99,6 @@ function todo.create_task(text, assignee)
...
-    local flow = table.parent.todo_add_button_flow
-    flow.todo_persist_button.destroy()
-
-    flow.add({
-        type = "button",
-        name = "todo_update_button_" .. index,
-        caption = {"todo.update"}
-    })
-end
-
```
Comment (hoylemd) [inline src/todo/todo.lua]: This was merged with the `create_add_edit` function in UI

**Comment by hoylemd** [comment]
Alrighty! Finally ready for review, whenever you have the time. I've also got a 'move to top/bottom' branch working, but it needs more polish before a PR.  Maybe later this week :)

File: src/todo/UI.lua (hunk 2 above)
Comment (JonasJurczok) [reply src/todo/UI.lua]: Works for me :)

File: src/control.lua (hunk 1 above)
Comment (JonasJurczok) [reply src/control.lua]: This should not be in master. Because if it is there then you can run faketorio tests, but it also requires everyone to have faketorio installed. One of the features of faketorio is to generate these lines for you when you run the `test` command :)

File: src/todo/UI.lua (hunk 3 above)
Comment (JonasJurczok) [reply src/todo/UI.lua]: I'm also new to this :) If it works fine. I think the usage of and and or is extremely unintuitive, though.

**Comment by JonasJurczok** [comment]
Awesome :)

Can you please report the problems with faketorio [here](https://github.com/JonasJurczok/faketorio/issues)? Then I'm aware of them and can try to reproduce/fix them directly.

File: locale/de/locale.cfg [hunk 7]
```diff
@@ -23,13 +24,13 @@ hide_done=Verberge abgeschlossene Aufgaben
 todolist-toggle-ui=Todo Liste minimieren/maximieren
 
 [mod-setting-name]
-todolist-show-minimized=Zeige "Todo List" Schalter wenn minimiert.
+todolist-show-button=Zeige "Todo List" Schalter ins der mod ui.
```
Comment (JonasJurczok) [inline locale/de/locale.cfg]: I suggest
`Zeige "Todo List" Schalter an/verstecke ihn`

File: spec/UI_feature.lua [hunk 8]
```diff
@@ -20,8 +20,8 @@ feature("Testing the UI", function()
...
 
-        assert(button == nil, "Maximize button found!")
+        assert(button ~= nil, "Maximize button not found!")
 
-        todo.show_minimized = temp
+        todo.show_button = temp
     end)
-end)
\ No newline at end of file
+end)
```
Comment (JonasJurczok) [inline spec/UI_feature.lua]: looks good.

File: src/control.lua [hunk 9]
```diff
@@ -22,16 +22,15 @@ end)
 
 script.on_event("todolist-toggle-ui", function(event)
     local player = game.players[event.player_index]
-    if todo.get_main_frame(player) then
-        todo.minimize(player)
-    else
-        todo.maximize(player)
-        todo.refresh_task_table(player)
-    end
+    todo.toggle_ui(player)
```
Comment (JonasJurczok) [inline src/control.lua]: Nice

File: src/control.lua (hunk 1 above)
Comment (JonasJurczok) [reply src/control.lua]: As already said in the main thread: This should not be here otherwise it won't work without faketorio anymore. Please remove it :D

File: src/todo/UI.lua (hunk 4 above)
Comment (JonasJurczok) [reply src/todo/UI.lua]: I did an experiment once and it should not be necessary, but I cannot recall the details anymore. So I aggree.. better safe than sorry :D

File: src/todo/UI.lua (hunk 5 above)
Comment (JonasJurczok) [reply src/todo/UI.lua]: yes

File: src/todo/style.lua [hunk 10]
```diff
@@ -0,0 +1,45 @@
+local default_gui = data.raw["gui-style"].default
```
Comment (JonasJurczok) [inline src/todo/style.lua]: Nice. I never got around to learn how styles work in Factorio. Thanks for implementing this :)

File: src/todo/todo.lua (hunk 6 above)
Comment (JonasJurczok) [reply src/todo/todo.lua]: I'm always a fan of deleting code :)

File: src/todo/todo.lua [hunk 11]
```diff
@@ -275,23 +251,23 @@ function todo.on_gui_click(event)
         local id = todo.get_task_id_from_element_name(element.name, "todo_item_down_")
         todo.move_down(id)
         todo.update_task_table()
-    else
-        todo.log("Unknown element name:" .. element.name)
+    elseif (string.find(element.name, "todo_")) then
+        todo.log("Unknown todo element name:" .. element.name)
```
Comment (JonasJurczok) [inline src/todo/todo.lua]: +1

File: src/control.lua (hunk 1 above)
Comment (hoylemd) [reply src/control.lua]: Ah! that explains why the lines suddenly appeared. Thanks, will do.

File: src/todo/UI.lua (hunk 3 above)
Comment (hoylemd) [reply src/todo/UI.lua]: Yeah, that's the downside of this pattern.  I'm thinking about just putting it in an if block anyways.  This is really nice and concise, but readability is more important.  I'll see what looks better.

File: src/todo/style.lua (hunk 10 above)
Comment (hoylemd) [reply src/todo/style.lua]: Too many years of front-end web development, finally paying off!

File: locale/de/locale.cfg (hunk 7 above)
Comment (hoylemd) [reply locale/de/locale.cfg]: If I'm reading that correctly (with my rather poor german augmented by google translate), it seems to translate to 'Show/Hide "Todo List" button'.  Maybe we just make them all say their equivalent of 'Show "Todo List" button', without mentioning hiding or the location.

I'm thinking that, because it's a checkbox/boolean value, the label should be descriptive of the 'on' state, with the state of the checkbox providing the polarity(?) of the setting (as in checked = 'this statement is true', unchecked = 'negate this statement')

File: src/control.lua (hunk 1 above)
Comment (hoylemd) [reply src/control.lua]: Actually, it does work without faketorio (at least on windows).  I'm impatient, so I packaged it all up and put the .zip on my windows gaming machine (where I don't have faketorio installed) and it worked great for a few hours last night.  Just FYI - still planning to remove it.

**Comment by hoylemd** [comment]
The final commit is just my suggestion for a more concise wording of the german option.  If @JonasJurczok Would prefer the show/hide phrasing, then I'll update all 3 roman-script localizations to the best of my ability (my french is better than my german. Yay Canadian education system!).  Someone who knows chinese will have to do that one though.

As for the faketorio issues, I'll post an issue, but It probably won't be super detailed.  I don't remember a lot of the specifics.

**Comment by Tarrke** [comment]
Please no Canadian French. Please please please. Promise I won't translate to English.

**joke intended**

**Comment by hoylemd** [comment]
@Tarrke  Oh, don't worry, the Ontario(my province) curriculum actually does teach *real* french, not the Quebecois pidgin :p

But I'm far from fluent - please do correct my bad french :p

(also thank you for specifying intent - I probably would have misinterpreted otherwise. )

**Comment by hoylemd** [comment]
Huh, I don't know what's up with that CI error.  Maybe it just needs to be run again?  Couldn't find a button for that :/

**Comment by hoylemd** [comment]
hoorj it worked!

**Comment by hoylemd** [comment]
@JonasJurczok Hey, I don't mean to be a bother, but do you think you could re-review the PR today or tomorrow?

**Comment by JonasJurczok** [comment]
Yes, sorry.. I got distracted with the faketorio issues on my windows
machine (development on windows is..........), Also I was barely at home
the last days.

I'll try to get it done tomorrow, otherwise Saturday.

Thanks for your work :)

Michael D. Hoyle <notifications@github.com> schrieb am Do., 9. Aug. 2018,
18:37:

**Comment by hoylemd** [comment]
:+1: Totally cool.  I feel you on the windows development. I have to use Azure a lot for work and it's..............

**Comment by hoylemd** [comment]
hoylemd is my handle!  But if you prefer real names, Mike Hoyle is good too.

Thanks! Now to rebase and PR my 'move to top/bottom' branch :p

**Comment by hoylemd** [comment]
Huh, the Travis build for the merge failed with that `busted could not be found` error again. I saw that while working on this branch. I wonder what's causing that?

**Comment by JonasJurczok** [comment]
I assume network error on Travis side. I planned to add all the luarocks
directory to the cache anyway. This would prevent it from downloading
everything again and again. Overall it should be fine though as the pr also
builds the merge result.

**Comment by hoylemd** [comment]
Yeah, I knew about the merge bit.  Just a little annoying.  Adding packages to the cache is a good idea though!
